### PR TITLE
chore: refine split enumerator for pulsar connector

### PR DESCRIPTION
**Pull request by shanicky**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

This PR fixes an issue with Pulsar split enumerator regarding non-partitioned-topic

The previous method to determine if a topic is partitioned was based on the `-partition-` suffix, which was wrong, now the `partitions` field in the metadata is used to determine.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/connector/src/source/pulsar/mod.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ pub use split::*;
 
 pub const PULSAR_CONNECTOR: &str = "pulsar";
 
-#[derive(Clone, Debug, Deserialize)]
+#[derive(Clone, Debug, Deserialize, Default)]
```
Comment (tabVersion) [inline src/connector/src/source/pulsar/mod.rs:27]: I think it is better to impl `Default` trait manually for clarity

**Review by tabVersion — APPROVED** [review]
rest LGTM

please don't work extra hours

File: src/connector/src/source/pulsar/admin/client.rs [hunk 2]
```diff
@@ -38,13 +39,20 @@ impl PulsarAdminClient {
     }
 
     pub async fn get_topic_metadata(&self, topic: &Topic) -> Result<PartitionedTopicMetadata> {
-        self.get(topic, "partitions").await
+        let res = self.http_get(topic, "partitions").await?;
+
+        if res.status() == StatusCode::NOT_FOUND {
+            return Err(anyhow!(
```
Comment (neverchanje) [inline src/connector/src/source/pulsar/admin/client.rs:45]: bail!()

File: src/connector/src/source/pulsar/enumerator/client.rs [hunk 3]
```diff
@@ -80,56 +78,52 @@ impl SplitEnumerator for PulsarSplitEnumerator {
         let offset = self.start_offset.clone();
         // MessageId is only used when recovering from a State
         assert!(!matches!(offset, PulsarEnumeratorOffset::MessageId(_)));
-        match self.topic.partition_index {
+
+        let topic_metadata = self.admin_client.get_topic_metadata(&self.topic).await?;
+        // note: may check topic exists by get stats
+        if topic_metadata.partitions < 0 {
```
[the remark this answers is not in the document]
Comment (neverchanje) [reply src/connector/src/source/pulsar/enumerator/client.rs:84]: I agree it's nearly impossible but having it still makes sense to me

**Review by neverchanje — APPROVED** [review]

**Review by yezizp2012 — APPROVED** [review]
Generally LGTM, you're so juan.

File: src/connector/src/source/pulsar/mod.rs (hunk 1 above)
Comment (shanicky) [reply src/connector/src/source/pulsar/mod.rs:27]: After thinking about it, it is confusing to give this kind of implementation Default, so I removed the Default

**Pull request merged**